File: public/java/042-clone-graph.java
/*
 * LeetCode Problem #133: Clone Graph
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/clone-graph/
 */
import java.util.*;

public class Solution {

    static class Node {
        int val;
        List<Node> neighbors;
        Node(int v) { val = v; neighbors = new ArrayList<>(); }
    }

    // APPROACH 1: DFS with Hash Map | O(N+E) time | O(N) space
    // EXPLAIN: Recursively clone each node, storing original→clone in a map to handle cycles.
    private Map<Node, Node> visited = new HashMap<>();

    public Node cloneGraph(Node node) {
        if (node == null) return null;
        if (visited.containsKey(node)) return visited.get(node);
        Node clone = new Node(node.val);
        visited.put(node, clone);
        for (Node nb : node.neighbors) clone.neighbors.add(cloneGraph(nb));
        return clone;
    }

    // APPROACH 2: BFS with Hash Map | O(N+E) time | O(N) space
    // EXPLAIN: BFS traversal; clone each node on first encounter, then wire neighbor edges.
    public Node cloneGraphBFS(Node node) {
        if (node == null) return null;
        Map<Node, Node> map = new HashMap<>();
        Queue<Node> q = new LinkedList<>();
        map.put(node, new Node(node.val));
        q.offer(node);
        while (!q.isEmpty()) {
            Node curr = q.poll();
            for (Node nb : curr.neighbors) {
                if (!map.containsKey(nb)) { map.put(nb, new Node(nb.val)); q.offer(nb); }
                map.get(curr).neighbors.add(map.get(nb));
            }
        }
        return map.get(node);
    }

    // APPROACH 3: DFS Iterative (Stack) | O(N+E) time | O(N) space
    // EXPLAIN: Explicit stack DFS; clone nodes on first visit, then add cloned neighbors.
    public Node cloneGraphStack(Node node) {
        if (node == null) return null;
        Map<Node, Node> map = new HashMap<>();
        Deque<Node> stack = new ArrayDeque<>();
        map.put(node, new Node(node.val));
        stack.push(node);
        while (!stack.isEmpty()) {
            Node curr = stack.pop();
            for (Node nb : curr.neighbors) {
                if (!map.containsKey(nb)) { map.put(nb, new Node(nb.val)); stack.push(nb); }
                map.get(curr).neighbors.add(map.get(nb));
            }
        }
        return map.get(node);
    }

    // APPROACH 4: Two-Pass BFS | O(N+E) time | O(N) space
    // EXPLAIN: First BFS creates all clones; second pass wires the neighbor edges.
    public Node cloneGraphTwoPass(Node node) {
        if (node == null) return null;
        Map<Node, Node> map = new HashMap<>();
        Queue<Node> q = new LinkedList<>();
        map.put(node, new Node(node.val));
        q.offer(node);
        while (!q.isEmpty()) {
            Node curr = q.poll();
            for (Node nb : curr.neighbors)
                if (!map.containsKey(nb)) { map.put(nb, new Node(nb.val)); q.offer(nb); }
        }
        for (Map.Entry<Node, Node> e : map.entrySet())
            for (Node nb : e.getKey().neighbors) e.getValue().neighbors.add(map.get(nb));
        return map.get(node);
    }

    // APPROACH 5: Recursive with Visited Set | O(N+E) time | O(N) space
    // EXPLAIN: Separate visited set and clone map; check visited before recursing into neighbors.
    private Map<Node, Node> clones = new HashMap<>();
    private Set<Node> seen = new HashSet<>();

    public Node cloneGraphVisitedSet(Node node) {
        if (node == null) return null;
        if (clones.containsKey(node)) return clones.get(node);
        Node clone = new Node(node.val);
        clones.put(node, clone);
        seen.add(node);
        for (Node nb : node.neighbors)
            clone.neighbors.add(seen.contains(nb) ? clones.get(nb) : cloneGraphVisitedSet(nb));
        return clone;
    }
}

// Made with Bob

File: public/python/042-clone-graph.py
# LeetCode Problem #133: Clone Graph
# Difficulty: Medium
# Link: https://leetcode.com/problems/clone-graph/

from typing import Optional
from collections import deque


class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


# ─────────────────────────────────────────────
# APPROACH 1: DFS with Hash Map | O(N+E) time | O(N) space
# EXPLAIN: Recursively clone each node, storing original→clone in a dict to handle cycles.
# WHEN: Natural recursive formulation; concise and easy to trace.

def cloneGraph_dfs(node: Optional[Node]) -> Optional[Node]:
    if not node:
        return None
    visited: dict = {}

    def dfs(n):
        if n in visited:
            return visited[n]
        clone = Node(n.val)
        visited[n] = clone
        for nb in n.neighbors:
            clone.neighbors.append(dfs(nb))
        return clone

    return dfs(node)


# ─────────────────────────────────────────────
# APPROACH 2: BFS with Hash Map | O(N+E) time | O(N) space
# EXPLAIN: BFS; clone each node on first encounter, then wire neighbor edges.
# WHEN: Iterative; avoids recursion depth issues on large graphs.

def cloneGraph_bfs(node: Optional[Node]) -> Optional[Node]:
    if not node:
        return None
    visited: dict = {node: Node(node.val)}
    q = deque([node])
    while q:
        curr = q.popleft()
        for nb in curr.neighbors:
            if nb not in visited:
                visited[nb] = Node(nb.val)
                q.append(nb)
            visited[curr].neighbors.append(visited[nb])
    return visited[node]


# ─────────────────────────────────────────────
# APPROACH 3: Iterative DFS (Stack) | O(N+E) time | O(N) space
# EXPLAIN: Explicit stack replaces recursion for DFS; same logic as Approach 1.
# WHEN: Avoids Python recursion limit for large or deeply connected graphs.

def cloneGraph_stack(node: Optional[Node]) -> Optional[Node]:
    if not node:
        return None
    visited: dict = {node: Node(node.val)}
    stack = [node]
    while stack:
        curr = stack.pop()
        for nb in curr.neighbors:
            if nb not in visited:
                visited[nb] = Node(nb.val)
                stack.append(nb)
            visited[curr].neighbors.append(visited[nb])
    return visited[node]


# ─────────────────────────────────────────────
# APPROACH 4: Two-Pass BFS | O(N+E) time | O(N) space
# EXPLAIN: First BFS creates all clones; second pass wires neighbor edges.
# WHEN: Cleaner separation of "create nodes" and "connect nodes" phases.

def cloneGraph_two_pass(node: Optional[Node]) -> Optional[Node]:
    if not node:
        return None
    visited: dict = {}
    q = deque([node])
    visited[node] = Node(node.val)
    while q:
        curr = q.popleft()
        for nb in curr.neighbors:
            if nb not in visited:
                visited[nb] = Node(nb.val)
                q.append(nb)
    for orig, clone in visited.items():
        for nb in orig.neighbors:
            clone.neighbors.append(visited[nb])
    return visited[node]


# ─────────────────────────────────────────────
# APPROACH 5: Recursive with Visited Set | O(N+E) time | O(N) space
# EXPLAIN: Maintain separate visited set alongside clone dict; check visited before recursing.
# WHEN: Explicit visited tracking makes the cycle-handling logic more transparent.

def cloneGraph_visited_set(node: Optional[Node]) -> Optional[Node]:
    if not node:
        return None
    clones: dict = {}
    seen: set = set()

    def dfs(n):
        if n in clones:
            return clones[n]
        clone = Node(n.val)
        clones[n] = clone
        seen.add(n)
        for nb in n.neighbors:
            if nb not in seen:
                clone.neighbors.append(dfs(nb))
            else:
                clone.neighbors.append(clones[nb])
        return clone

    return dfs(node)


# Made with Bob
